### PR TITLE
8309841: Jarsigner should print a warning if an entry is removed

### PR DESCRIPTION
**Pull request by wangweij**
There ~are two~ is one change~s~:

1. In `jarsigner -verify`, check a .SF file contains un-existing entries and print them out as
```
Warning: nonexistent signed entries detected: [a]
```
~2. In `JarSigner::sign0`, when creating a new .SF file, only include signed file entries.~

*Update*: Even when the JAR file is re-signed, the hash entry for the missing file will be in the new .SF file. There is no way to tell if this is for a file entry or a user-defined entry.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8334261](https://bugs.openjdk.org/browse/JDK-8334261) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8309841](https://bugs.openjdk.org/browse/JDK-8309841): Jarsigner should print a warning if an entry is removed (**Enhancement** - P3)
 * [JDK-8334261](https://bugs.openjdk.org/browse/JDK-8334261): Jarsigner should print a warning if an entry is removed (**CSR**)

### Reviewers
 * [Hai-May Chao](https://openjdk.org/census#hchao) (@haimaychao - Committer) Review applies to [8311e890](https://git.openjdk.org/jdk/pull/19599/files/8311e8902eec28d15b8c6ce6bbd7e252b142785b)
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19599/head:pull/19599` \
`$ git checkout pull/19599`

Update a local copy of the PR: \
`$ git checkout pull/19599` \
`$ git pull https://git.openjdk.org/jdk.git pull/19599/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19599`

View PR using the GUI difftool: \
`$ git pr show -t 19599`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19599.diff">https://git.openjdk.org/jdk/pull/19599.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19599#issuecomment-2346315420)

### CODE REVIEW

**Comment by wangweij** [comment]
Ask again for review.

File: test/jdk/sun/security/tools/jarsigner/RemovedFiles.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                Files.writeString(Path.of("b"), "b"));
+        SecurityTools.keytool("-genkeypair -storepass changeit -keystore ks -alias x -dname CN=x -keyalg ed25519");
+        SecurityTools.jarsigner("-storepass changeit -keystore ks a.jar x");
+
+        // All is fine at the beginning.
+        SecurityTools.jarsigner("-verify a.jar")
+                .shouldNotContain("Nonexistent signed entries detected. See details in -verbose output.");
+
+        // Remove an entry after signing. There will be a warning.
+        JarUtils.deleteEntries(Path.of("a.jar"), "a");
```
[the remark this answers is not in the document]
Comment (wangweij) [reply test/jdk/sun/security/tools/jarsigner/RemovedFiles.java:52]: I can imagine some info on whether the entry is really removed or not could be useful for someone, I'll make an update to the `deleteEntries` method. But, I don't want to care about it here. The next lines will confirm the removal.

**Comment by haimaychao** [comment]
CSR looks good to me. Clarifying that non-file entries in MANIFEST.MF could cause false positive is useful in the CSR.

File: src/jdk.jartool/share/classes/sun/security/tools/jarsigner/Resources.java [hunk 2]
```diff
@@ -164,6 +164,7 @@ public class Resources extends java.util.ListResourceBundle {
 
         {"history.with.ts", "- Signed by \"%1$s\"\n    Digest algorithm: %2$s\n    Signature algorithm: %3$s, %4$s\n  Timestamped by \"%6$s\" on %5$tc\n    Timestamp digest algorithm: %7$s\n    Timestamp signature algorithm: %8$s, %9$s"},
         {"history.without.ts", "- Signed by \"%1$s\"\n    Digest algorithm: %2$s\n    Signature algorithm: %3$s, %4$s"},
+        {"history.nonexistent.entries", "    Warning: nonexistent signed entries: "},
```
Comment (nibjen) [inline src/jdk.jartool/share/classes/sun/security/tools/jarsigner/Resources.java:167]: Inline with existing warning message, can it be "WARNING: Nonexistent signed entries: "

Comment (wangweij) [reply src/jdk.jartool/share/classes/sun/security/tools/jarsigner/Resources.java:167]: This warning is a part of a block on signer info and I'd like the word does not stand out too much. For example, the test shows:
```
- Signed by "CN=x"
    Digest algorithm: SHA-384
    Signature algorithm: Ed25519, 255-bit key
    Warning: nonexistent signed entries: [a]
```

**Comment by haimaychao** [comment]
Suggest to add testcases to remove more than one entry or file ending with "*" in RemovedFiles.

**Comment by haimaychao** [comment]
Do you think it may be useful to have a testcase that falsely reports the missing entry?

**Comment by wangweij** [comment]
> Suggest to add testcases to remove more than one entry or file ending with "*" in RemovedFiles.
This can be a test to the `JarUtils` method itself. I can add one.
> Do you think it may be useful to have a testcase that falsely reports the missing entry?
I'll add one.

**Comment by haimaychao** [comment]
Thanks for adding more testcases for this PR. Changes look good.

**Review by haimaychao — APPROVED** [review]

File: src/jdk.jartool/share/classes/sun/security/tools/jarsigner/Resources.java [hunk 3]
```diff
@@ -178,6 +179,7 @@ public class Resources extends java.util.ListResourceBundle {
         {"key.bit.disabled", "%d-bit key (disabled)"},
         {"key.bit.eccurve.disabled", "%1$d-bit %2$s key (disabled)"},
         {"unknown.size", "unknown size"},
+        {"nonexistent.entries.found", "Nonexistent signed entries detected. See details in -verbose output."},
```
Comment (seanjmullan) [inline src/jdk.jartool/share/classes/sun/security/tools/jarsigner/Resources.java:182]: For the second sentence, in other warning messages, we say "Re-run jarsigner with the -verbose option for more details." Perhaps we should be consistent?

Comment (seanjmullan) [reply src/jdk.jartool/share/classes/sun/security/tools/jarsigner/Resources.java:182]: Actually, I see the output will also contain the message "Re-run with the -verbose and -certs options for more details." so I take back my comment above.

Comment (seanjmullan) [reply src/jdk.jartool/share/classes/sun/security/tools/jarsigner/Resources.java:182]: For this summary message, I suggest we be a bit more descriptive like in other messages:

"This jar contains signed entries for files that do not exist. See the -verbose output for more details."

Comment (wangweij) [reply src/jdk.jartool/share/classes/sun/security/tools/jarsigner/Resources.java]: Updated. Thanks.

File: test/lib-test/jdk/test/lib/util/JarUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                Files.writeString(Path.of("c"), ""));
+        Asserts.assertEquals(Set.of("a", "b1", "b2", "bx/x", "c"), content("a.jar"));
+
+        JarUtils.deleteEntries(Path.of("a.jar"), "a");
+        Asserts.assertEquals(Set.of("b1", "b2", "bx/x", "c"), content("a.jar"));
+
+        // Note: b* covers everything starting with b, even bx/x
+        JarUtils.deleteEntries(Path.of("a.jar"), "b*");
+        Asserts.assertEquals(Set.of("c"), content("a.jar"));
+    }
```
Comment (seanjmullan) [inline test/lib-test/jdk/test/lib/util/JarUtilsTest.java:59]: You could also add a test where the pattern doesn't match, and a test where there is more than one pattern, and the matching pattern is not the first one.

Comment (wangweij) [reply test/lib-test/jdk/test/lib/util/JarUtilsTest.java:69]: Updated.

**Review by seanjmullan — APPROVED** [review]

**Pull request closed without merging**